c_src/myhtml_worker.c

```diff
@@ -256,140 +257,134 @@ read_parse_flags(ETERM* list)
+      tstack_push(&stack, erl_mk_empty_list());
 
-    // attributes
-    attrs = build_node_attrs(prefab, tree, node);
+      prev_node = current_node;
+      current_node=current_node->last_child;
+
+      continue;
+    } else {
+      if ((myhtml_node_is_close_self(current_node)  || myhtml_node_is_void_element(current_node))
+      && (*parse_flags & FLAG_NIL_SELF_CLOSING)) {
+        children = prefab->atom_nil;
+      } else {
+        children = prefab->empty_list;
+      }
+    }
 
-    // children
-    children = build_node_children(prefab, tree, node, parse_flags);
+    myhtml_tag_id_t tag_id = myhtml_node_tag_id(current_node);
+    myhtml_namespace_t tag_ns = myhtml_node_namespace(current_node);
 
-    if (!(*parse_flags & FLAG_HTML_ATOMS) || (tag_id == MyHTML_TAG__UNDEF || tag_id == MyHTML_TAG_LAST_ENTRY || tag_ns != MyHTML_NAMESPACE_HTML))
+    if (tag_id == MyHTML_TAG__TEXT)
     {
-      tag = erl_mk_binary(tag_string, tag_string_len);
-      /* ETERM* tuple3[] = {tag, attrs, children}; */
-      /* result = erl_mk_tuple(tuple3, 3); */
-      result = erl_format("{~w, ~w, ~w}", tag, attrs, children);
+      size_t text_len;
+
+      const char* node_text = myhtml_node_text(current_node, &text_len);
+      result = erl_mk_binary(node_text, text_len);
     }
-    else
+    else if (tag_id == MyHTML_TAG__COMMENT)
     {
-      // tag = erl_mk_atom(tag_string);
-      tag = erl_mk_atom(tag_string);
-      /* ETERM* tuple3[] = {tag, attrs, children}; */
-      /* result = erl_mk_tuple(tuple3, 3); */
-      result = erl_format("{~w, ~w, ~w}", tag, attrs, children);
+      size_t comment_len;
+      const char* node_comment = myhtml_node_text(current_node, &comment_len);
+
+      // For <!----> myhtml_node_text will return a null pointer, which will make erl_format segfault
+      if(comment_len == 0) {
+        node_comment= (char*) "";
+      }
+
+      ETERM* comment = erl_mk_binary(node_comment, comment_len);
+
+      if (*parse_flags & FLAG_COMMENT_TUPLE3)
+      {
+        result = erl_format("{comment, [], ~w}", comment);
+      }
+      else
+      {
+        result = erl_format("{comment, ~w}", comment);
+      }
     }
-
-    // free allocated resources
-    if (tag_ns != MyHTML_NAMESPACE_HTML)
+    else
     {
-      free(tag_ns_buffer);
+      ETERM* tag;
+      ETERM* attrs;
+
+      // get name of tag
+      size_t tag_name_len;
+      const char *tag_name = myhtml_tag_name_by_id(tree, tag_id, &tag_name_len);
+      // get namespace of tag
+      size_t tag_ns_len;
+      const char *tag_ns_name_ptr = myhtml_namespace_name_by_id(tag_ns, &tag_ns_len);
+      char *tag_ns_buffer;
+      char buffer [tag_ns_len + tag_name_len + 1];
+      char *tag_string = buffer;
+      size_t tag_string_len;
+
+      if (tag_ns != MyHTML_NAMESPACE_HTML)
+      {
+        // tag_ns_name_ptr is unmodifyable, copy it in our tag_ns_buffer to make it modifyable.
+        tag_ns_buffer = malloc(tag_ns_len);
```

Comment:
You're in a new scope, so you can use a VLA.  You also want to add a second byte for the nul terminator.

```
char tag_ns_buffer[tag_ns_len + tag_name_len + 2];
snprintf(tag_ns_buffer, sizeof tag_ns_buffer, "%s:%s", tag_ns_name_ptr, tag_name);
lowercase(tag_ns_buffer);
```